cost_reports/finops.py

```diff
@@ -0,0 +1,184 @@
+    by_catalog = no_center.groupby((lambda row_index: group_by_tag(no_center,row_index,SERVICE_CATALOG_PRINCIPAL_COLUMN,"Service Catalog Provisioner")), dropna=False, group_keys=False)
+
+    # sum up each category and sort, descending
+    center_summed = by_center.sum().sort_values(VALUE_COLUMN, ascending=False)[[VALUE_COLUMN]]
+    stack_summed = by_stack.sum().sort_values(VALUE_COLUMN, ascending=False)[[VALUE_COLUMN]]
+    catalog_summed = by_catalog.sum().sort_values(VALUE_COLUMN, ascending=False)[[VALUE_COLUMN]]
+
+    # display the cost centers
+    report_body += "\nCosts by Program\n"
+    sorted = center_summed.rename(columns={VALUE_COLUMN:DISPLAY_VALUE_COLUMN})
+    with pd.option_context( 'display.precision', 2),\
+        pd.option_context('display.max_rows', 500),\
+        pd.option_context('display.max_colwidth', 200),\
+        pd.option_context('display.float_format', (lambda x : f'${x:.2f}')):
+        report_body += sorted.to_string()
+
+    total = sorted[DISPLAY_VALUE_COLUMN].sum()
+    if (abs(total-total_cost)>0.01):
+        raise Exception("categorized costs do not add up to total bill.")
+    report_body += f"\nTotal: ${total:.2f}\n"
+
+    # display cfn stacks and service catalog owners for line items with no cost center
+    for summed in [stack_summed, catalog_summed]:
+        report_body += "\nCosts without a known Program\n"
+        sorted = summed[summed[VALUE_COLUMN] >= 0.01].rename(columns={VALUE_COLUMN:DISPLAY_VALUE_COLUMN})
+        with pd.option_context( 'display.precision', 2),\
+            pd.option_context('display.max_rows', 500),\
+            pd.option_context('display.max_colwidth', 200),\
+            pd.option_context('display.float_format', (lambda x : f'${x:.2f}')):
+            report_body += sorted.to_string()
+
+    return report_body
+
+def safe_at(df, r, c):
+    try:
+        return df.at[r,c]
+    except KeyError:
+        return None
+
+def valid_value(check_value):
+    if check_value is None or check_value == 'na':
+        return False
+    return True
+
+def group_by_tag(df, row_index, tag_column, heading):
+    tag = safe_at(df, row_index, tag_column)
+    if not valid_value(tag):
+        return f"{heading} / None"
+    return f"{heading} / {tag}"
+
+def cost_center_lookup(df, row_index):
+    cost_center = safe_at(df, row_index, COST_CENTER_COLUMN)
+
+    # don't try to process an invalid value
+    if not valid_value(cost_center):
+        return cost_center
+
+    # check secondary cost center for Other values
+    if cost_center == 'Other / 000001':
+        cost_center = safe_at(df, row_index, COST_CENTER_OTHER_COLUMN)
+        if not valid_value(cost_center):
+            cost_center = 'Other (Unspecified) / 000001'
+
+    # replace tags that look like "name_123456" with "name / 123456"
+    cost_center = re.sub(r"(_)([0-9]{5,6})", (lambda x: " / "+x.group(2)), cost_center)
+
+    # normalize 000000 case
+    cost_center = TAG_REPLACEMENT.get(cost_center, cost_center)
```

Comment:
i don't get why the default object is the same as what you are trying to get?

Reply:
If cost_center is a key in the TAG_REPLACEMENT map, then we will return the associated value for that key, but if it's not in the map then we want to keep the original value. I can make this a little more logical.